src/openms/source/ANALYSIS/OPENSWATH/MRMDecoy.cpp

```diff
@@ -329,50 +349,60 @@ namespace OpenMS
     // Go through all peptides and apply the decoy method to the sequence
     // (pseudo-reverse, reverse or shuffle). Then set the peptides and proteins of the decoy
     // experiment.
+    Size progress = 0;
+    startProgress(0, exp.getPeptides().size(), "Generating decoy peptides");
     for (Size pep_idx = 0; pep_idx < exp.getPeptides().size(); ++pep_idx)
     {
+      setProgress(++progress);
+
       OpenMS::TargetedExperiment::Peptide peptide = exp.getPeptides()[pep_idx];
-      // continue if the peptide has C/N terminal modifications and we should exclude them
-      if (remove_CNterminal_mods && MRMDecoy::has_CNterminal_mods(peptide)) {continue; }
+
       peptide.id = decoy_tag + peptide.id;
-      OpenMS::String original_sequence = peptide.sequence;
+
       if (!peptide.getPeptideGroupLabel().empty())
       {
         peptide.setPeptideGroupLabel(decoy_tag + peptide.getPeptideGroupLabel());
       }
 
       if (method == "pseudo-reverse")
       {
-        peptide = MRMDecoy::pseudoreversePeptide(peptide);
+        // exclude peptide if it has C/N terminal modifications because we can't do a (partial) reverse
+        if (MRMDecoy::has_CNterminal_mods(peptide))
+        {
+          LOG_DEBUG << "[peptide] Skipping " << peptide.id << " due to C/N-terminal modifications" << std::endl;
+          exclusion_peptides.push_back(peptide.id);
+        }
+        else
+        {
+          peptide = MRMDecoy::pseudoreversePeptide(peptide);
+        }
       }
       else if (method == "reverse")
       {
-        peptide = MRMDecoy::reversePeptide(peptide);
+        // exclude peptide if it has C/N terminal modifications because we can't do a (partial) reverse
+        if (MRMDecoy::has_CNterminal_mods(peptide))
+        {
+          LOG_DEBUG << "[peptide] Skipping " << peptide.id << " due to C/N-terminal modifications" << std::endl;
+          exclusion_peptides.push_back(peptide.id);
+        }
+        else
+        {
+          peptide = MRMDecoy::reversePeptide(peptide);
+        }
       }
       else if (method == "shuffle")
       {
         peptide = MRMDecoy::shufflePeptide(peptide, identity_threshold, -1, max_attempts);
       }
+
       for (Size prot_idx = 0; prot_idx < peptide.protein_refs.size(); ++prot_idx)
       {
         peptide.protein_refs[prot_idx] = decoy_tag + peptide.protein_refs[prot_idx];
       }
 
-      if (MRMDecoy::AASequenceIdentity(original_sequence, peptide.sequence) > identity_threshold)
```

Comment:
The amino acid sequence identity threshold is not a control anymore, but an optimization parameter for the ``shuffle`` method. The difference to the target peptides is directly controlled by comparing the fragment ion m/z.